tokenizers/src/tokenizer/mod.rs

```diff
@@ -845,22 +847,35 @@ where
 
     /// Decode the given ids, back to a String
     pub fn decode(&self, ids: &[u32], skip_special_tokens: bool) -> Result<String> {
-        let tokens = ids
-            .iter()
-            .filter_map(|id| {
-                self.added_vocabulary
-                    .id_to_token(*id, &self.model)
-                    .filter(|token| {
-                        !skip_special_tokens || !self.added_vocabulary.is_special_token(token)
-                    })
-            })
-            .collect::<Vec<_>>();
-
-        if let Some(decoder) = &self.decoder {
-            decoder.decode(tokens)
-        } else {
-            Ok(tokens.join(" "))
+        let mut result = String::with_capacity(ids.len());
+        let mut chunks = Vec::with_capacity(ids.len());
+        for id in ids {
+            if let Some(added_token) = self.added_vocabulary.simple_id_to_token(*id) {
+                if skip_special_tokens && self.added_vocabulary.is_special_token(&added_token) {
+                    continue;
+                }
+                let text_chunk = if let Some(decoder) = &self.decoder {
+                    decoder.decode(chunks.clone())?
+                } else {
+                    chunks.join(" ")
+                };
+                result.push_str(&text_chunk);
+                if !result.is_empty() {
+                    result.push(' ');
+                }
```

Comment:
why are we adding an extra space before the added token? IMO we should only do this if there is no decoder (default join on `" "`)

Reply:
https://github.com/huggingface/tokenizers/actions/runs/8819865853/job/24212114706#step:12:262

Reply:
These test are done when there is no `decoder` no? 
I mean the previous behaviour does not add spaces between added token if there is a decoder

Reply:
Yes it did (because the test wasn't using a decoder, so by default no decoder means space separated)